### PR TITLE
Provide `make_napari_viewer` as a pytest plugin, for external use.

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR would move some of our test fixtures (currently just make_test_viewer) into a new file (`napari.utils._testsupport`) to make it easier to export it as a pytest plugin at the `pytest11` entry_point.  This would make it easier for other packages to use that fixture (such as in napari/napari-console#7)

We might also consider renaming it to `make_napari_viewer` so that it is clearer where the fixture comes from when using externally.  We can also put additional fixtures in `_testsupport` if anyone wants

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
https://docs.pytest.org/en/stable/writing_plugins.html#making-your-plugin-installable-by-others

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
i'm confused by the codecov results here, saying 66.82% coverage.  When I run these tests with pytest-cov locally, I get no change in coverage:
```
statements | missing | excluded | coverage
33895      | 6880    | 21       | 80%
```

**Comment by sofroniewn** [comment]
this looks great @tlambert03. that is funny with codecov, but i wouldn't worry further about it.

I'd support renaming `make_napari_viewer` - curious what @jni thinks.

On the plugin side do I now need to do anything to discover these methods or with `pluggy` handle that for me?

**Review by sofroniewn — APPROVED** [review]

**Review by jni — APPROVED** [review]
I'm also in favour of make_napari_viewer, but anyway this LGTM!

**Comment by tlambert03** [comment]
> make_napari_viewer

can do quickly

**Comment by tlambert03** [comment]
> can do quickly

ugh... merged master and now I'm getting a leaked widget for some reason.  might not finish debugging tonight, but will pick it up tomorrow

**Comment by tlambert03** [comment]
> I'm getting a leaked widget for some reason

see #2161 ... it wasn't directly related to this PR (though it does need addressing).  But this is ready to go

**Comment by sofroniewn** [comment]
> see #2161 ... it wasn't directly related to this PR (though it does need addressing). But this is ready to go

ok can discuss there. Will merge now. After the next napari release will start using in `napari-console`

**Comment by tlambert03** [comment]
Important note for developers testing locally: you may need to run `pip install -e .` after merging this PR

**Comment by joshmoore** [comment]
Quick note that this ended up being breaking for "end-users" (i.e. ome-zarr-py). Bit of a surprise for a point release. As a side-note, in trying to get the GitHub actions to go green again, I'm currently running into this:

```
>           raise ValueError(f"no option named {name!r}") from e
E           ValueError: no option named '--show-viewer'

/usr/local/anaconda3/envs/z/lib/python3.9/site-packages/_pytest/config/__init__.py:1474: ValueError
======================================================= short test summary info =======================================================
FAILED tests/test_napari.py::TestNapari::test_viewer - ValueError: no option named '--show-viewer'
```

<details><summary>napari --info</summary>

```
(z) /opt/ome-zarr-py $napari --info
/usr/local/anaconda3/envs/z/lib/python3.9/site-packages/napari/__main__.py:260: UserWarning: pythonw executable not found.
To unfreeze the menubar on macOS, click away from napari to another app, then reactivate napari. To avoid this problem, please install python.app in conda using:
conda install -c conda-forge python.app
  warnings.warn(msg)
napari: 0.4.4
Platform: macOS-11.1-x86_64-i386-64bit
System: MacOS 11.1
Python: 3.9.1 | packaged by conda-forge | (default, Jan 26 2021, 01:32:59)  [Clang 11.0.1 ]
Qt: 5.12.9
PyQt5: 5.12.3
NumPy: 1.20.0
SciPy: 1.6.0
Dask: 2021.01.1
VisPy: 0.6.6

OpenGL:
  - GL version:  2.1 ATI-4.2.13
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen #1: resolution 1920x1080, scale 1.0
  - screen #2: resolution 1536x960, scale 2.0

Plugins:
  - svg: 0.1.4
 ```

</details>

**Comment by tlambert03** [comment]
So sorry @joshmoore, can you point me to your failing tests so I can have a look?

One thing I'm wondering about is your import of our conftest.py ... that wasn't ever intended as a public way to use our fixtures, (that's just an internal pytest config file) and that may be hiding the "actual" plugin fixture.  But happy to dig further with some logs

**Comment by tlambert03** [comment]
But the show viewer thing: that's our fault!  I really thought that pytest_adopts function made it into our _testsuport file :/ ... will need release a fix for that

**Comment by joshmoore** [comment]
> will need release a fix for that

Cool. Thanks for letting me know. I won't push my intermediate work then to https://github.com/ome/ome-zarr-py/pull/69. Let me know when you want me to try a follow on.

> (that's just an internal pytest config file)

Heh. Yeah. That was the only way I could get it to pass with @Czaki. I'm leaning towards extracting the napari work out of ome-zarr in favor of `napari-ome-zarr`. Thoughts?

**Comment by Czaki** [comment]
> So sorry @joshmoore, can you point me to your failing tests so I can have a look?
> 
> One thing I'm wondering about is your import of our conftest.py ... that wasn't ever intended as a public way to use our fixtures, (that's just an internal pytest config file) and that may be hiding the "actual" plugin fixture. But happy to dig further with some logs

To test napari based things there is a need to make a Viewer instance. Napari code structure is really dynamic the maintenance of its own method will be very painful. The best is to left all shutdown code on side of napari. 

Maybe `make_napari_viewer` should be implemented in the `utils` module and only registered in `conftest.py`. But it is really important to expose it to developers of the dependent packages. 

Registering napari as the `pytest` plugin causes the napari test to fail when such a package is installed.

**Comment by joshmoore** [comment]
I've pushed https://github.com/ome/ome-zarr-py/pull/69/commits/9731606f10ceb379d9d0628553a096bcb74dc69c but it will fail due to https://github.com/napari/napari/pull/2131#issuecomment-773228885 (unless @tlambert03 is faster with 0.4.5)

**Comment by tlambert03** [comment]
> But it is really important to expose it to developers of the dependent packages.

@czaki, the proper way to expose fixtures to outside packages is to use the pytest plugin system exporting your modules that include pytest fixtures as a [`pytest11` entry_point](https://docs.pytest.org/en/stable/writing_plugins.html#making-your-plugin-installable-by-others), as is done in this PR. The pytest docs themselves state that conftest files are [local per-directory configs](https://docs.pytest.org/en/stable/writing_plugins.html#plugin-discovery-order-at-tool-startup) and not meant to be used elsewhere.

So to summarize:
- no one should use our conftest.py outside of napari.
- napari.utils._testsupport is our "pytest plugin" and other packages should only use stuff there.
- this PR had an omission and forgot to include the `pytest_add_opts` that adds the `--show-viewer` options, which @joshmoore mentioned above.  that will be fixed soon.

**Comment by Czaki** [comment]
@joshmoore Maybe it will be nice to add --pre tests 
https://github.com/napari/napari/blob/master/.github/workflows/test_prereleases.yml

Then You will be automatically warned when pre-release is available and have time to protest.

**Comment by tlambert03** [comment]
> (unless @tlambert03 is faster with 0.4.5)

@joshmoore, I'll fix this today, and will try to figure out a way to add a fixture test to help prevent this.  I _think_ we could do a 0.4.5 soon, but @sofroniewn and @jni can weigh in on whether there is anything already in master that needs to be extracted first

**Comment by tlambert03** [comment]
added a fix and a "test test" :) in #2208

> I'm leaning towards extracting the napari work out of ome-zarr in favor of napari-ome-zarr. Thoughts?

that's your call.  While we'll definitely _try_ not to ever break your stuff as we develop, we can't guarantee it, so if that decoupling makes you more comfortable, I get that.  (in this _particular_ case, I kinda feel like it wasn't a public API that broke, but rather that you had been encouraged to use an internal feature in an way that wasn't explicitly supported ...)

**Comment by sofroniewn** [comment]
> @joshmoore, I'll fix this today, and will try to figure out a way to add a fixture test to help prevent this. I think we could do a 0.4.5 soon, but @sofroniewn and @jni can weigh in on whether there is anything already in master that needs to be extracted first

I think we can get release out soon, but it would be good to get from @joshmoore a level of urgency - is tonight needed, or early next week ok?

> Quick note that this ended up being breaking for "end-users" (i.e. ome-zarr-py). Bit of a surprise for a point release. As a side-note, in trying to get the GitHub actions to go green again, I'm currently running into this:

It's not quite clear to be are just the tests broken or is running/ installing the package broken too.

If just the tests, maybe we try and make the release early next week and apologize for the broken CI in the mean time. If the whole package is broken maybe we can try and do it faster. I'd rather not pull anything out of what we've already merged though.

**Comment by tlambert03** [comment]
> It's not quite clear to be are just the tests broken or is running/ installing the package broken too.

I believe it's just the tests... and in this case, it's also because they were importing our conftest.py file directly.  I've spoken with @joshmoore and we're trying to get them setup with the workaround I mentioned, and off using our conftest.py

**Comment by Czaki** [comment]
> @Czaki, the proper way to expose fixtures to outside packages is to use the pytest plugin system exporting your modules that include pytest fixtures as a [`pytest11` entry_point](https://docs.pytest.org/en/stable/writing_plugins.html#making-your-plugin-installable-by-others), as is done in this PR. The pytest docs themselves state that conftest files are [local per-directory configs](https://docs.pytest.org/en/stable/writing_plugins.html#plugin-discovery-order-at-tool-startup) and not meant to be used elsewhere.

@tlambert03 try of registering napari as pytest plugin in `ome-zarr-py` breaks napari test earlier. Import from conftest is a workaround for this. in #2208 I do not see registering napari as a pytest plugin.

**Comment by tlambert03** [comment]
>  Import from conftest is a workaround

I know you were just being helpful, but it's not a very good workaround.  As pytest's docs mention, `conftest.py` is _local_.  nobody should be importing other packages' conftest.py and it's not part of our public offerings.  (The only reason we include ours in source is so that we can easily test our sdist).  I would encourage people to _vendor_ the patterns of our conftest.py that they found useful (until, we offer a "real" pytest plugin... which we now do)

> I do not see registering napari as a pytest plugin.

It's already in master, here:

https://github.com/napari/napari/blob/79ab3589eb398ce56d2d0a16025cdb7f435ffaa5/setup.cfg#L115-L116

**Comment by joshmoore** [comment]
> I think we can get release out soon, but it would be good to get from @joshmoore a level of urgency - is tonight needed, or early next week ok?

I was very much asleep, so "tonight" was definitely not necessary. :wink: With @tlambert03's workaround, https://github.com/ome/ome-zarr-py/pull/69 has gone green, so no one is blocked from my side.
